src/Raven.Server/Documents/Queries/Graph/GraphQueryPlan.cs

```diff
@@ -196,5 +199,87 @@ public List<Match> Execute()
+            indexNamesGatherer.Visit(_rootQueryStep);
+            var indexes = new List<Index>();
+            Dictionary<Index, AsyncWaitForIndexing> indexWaiters = new Dictionary<Index, AsyncWaitForIndexing>();
+            var queryTimeout = _query.WaitForNonStaleResultsTimeout ?? Index.DefaultWaitForNonStaleResultsTimeout;
+            foreach (var indexName in indexNamesGatherer.Indexes)
+            {
+                var index = _database.IndexStore.GetIndex(indexName);
+                indexes.Add(index);
+                indexWaiters.Add(index, new AsyncWaitForIndexing(queryDuration, queryTimeout, index));
+            }
+
+            var staleIndexes = indexes;
+            while (true)
+            {
+                //If we found a stale index we already disposed of the old transaction
+                if (_context.Transaction == null || _context.Transaction.Disposed)
+                    _context.OpenReadTransaction();
+
+                //see http://issues.hibernatingrhinos.com/issue/RavenDB-5576
+                var frozenAwaiters = new Dictionary<Index, AsyncManualResetEvent.FrozenAwaiter>();
+                foreach (var index in staleIndexes)
+                {
+                    frozenAwaiters.Add(index, index.GetIndexingBatchAwaiter());
+                }
+                staleIndexes = GetStaleIndexes(staleIndexes, etag);
+                //All indexes are not stale we can continue without waiting
+                if (staleIndexes.Count == 0)
+                {
+                    //false means we are not stale
+                    return false;
+                }
+
+                bool foundStaleIndex = false;
+                bool indexTimedout = false;
+                //here we will just wait for the first stale index we find
+                foreach (var index in staleIndexes)
+                {
+                    var indexAwaiter = indexWaiters[index];
+                    //if any index timedout we are stale
+                    indexTimedout |= indexAwaiter.TimeoutExceeded;
+
+                    if (Index.WillResultBeAcceptable(true, _query, indexWaiters[index]) == false)
+                    {
+                        _context.CloseTransaction();                        
+                        await indexAwaiter.WaitForIndexingAsync(frozenAwaiters[index]).ConfigureAwait(false);
+                        foundStaleIndex = true;                        
+                        break;
+                    }
+                }
+
+                //We are done waiting for all stale indexes
+                if (foundStaleIndex == false)
+                {
+                    //we might get here if all indexes have timedout 
+                    return indexTimedout;
+                }
+            }
+        }
+
+        private List<Index> GetStaleIndexes(List<Index> indexes, long etag)
+        {
+            var staleList = new List<Index>();
```

Comment:
Why not re-use the same instance? I don't see it being used for other stuff?

Reply:
I can't modify the list while iterating on it, so I need a to delete list anyway so I might as well just create a new one.